pkg/controller/direct/monitoring/refs.go

```diff
@@ -85,10 +89,76 @@ func normalizeResourceName(ctx context.Context, reader client.Reader, src client
 	return ref, nil
 }
 
+func normalizeMonitoringAlertPolicyRef(ctx context.Context, reader client.Reader, src client.Object, project references.Project, ref *krm.MonitoringAlertPolicyRef) (*krm.MonitoringAlertPolicyRef, error) {
+	if ref == nil {
+		return nil, nil
+	}
+
+	if ref.Name == "" && ref.External == "" {
+		return nil, fmt.Errorf("must specify either name or external on reference")
+	}
+	if ref.Name != "" && ref.External != "" {
```

Comment:
I saw the CRD already use oneOf.
But anyways, I think this entire function is helpful for all ResRef.

Reply:
Right.  It doesn't cost us much, and then we don't need to rely on CRD validation.

BTW, I'm thinking we can greatly reduce code duplication here in future, maybe using AdapterForURL, or maybe just by adding a method to the Adapter called ResolveReference or something.  But ... not a problem right now.

Reply:
Yes! I like this idea